File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/methodadapter/core/CoreMixinPatternMatcher.java
package net.yakclient.mixin.base.internal.methodadapter.core;

import net.yakclient.mixin.base.YakMixins;
import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils;
import net.yakclient.mixin.base.internal.instruction.Instruction;
import net.yakclient.mixin.base.internal.instruction.core.CoreInsnExecutor;
import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher;
import org.objectweb.asm.MethodVisitor;

public abstract class CoreMixinPatternMatcher extends MethodVisitor implements MixinPatternMatcher {
    static final int NOT_MATCHED = 0;
    private final Instruction instructions;
    int state = NOT_MATCHED;
    private int locals = 1;

    public CoreMixinPatternMatcher(MethodVisitor visitor, Instruction instructions) {
        super(YakMixins.ASM_VERSION, visitor);
        this.instructions = instructions;
    }


    void executeInsn() {
        this.executeInsn(this.instructions);
    }

    private void executeInsn(Instruction insn) {
        if (this.mv instanceof CoreMixinPatternMatcher) ((CoreMixinPatternMatcher) this.mv).executeInsn(insn);
        else new CoreInsnExecutor(insn, this.mv).execute();
    }

    @Override
    public void visitInsn(int opcode) {
        this.visitInsn();
        super.visitInsn(opcode);
    }

    @Override
    public void visitIntInsn(int opcode, int operand) {
        this.visitInsn();
        super.visitIntInsn(opcode, operand);
    }

    @Override
    public void visitVarInsn(int opcode, int var) {
        if (this.mv instanceof CoreMixinPatternMatcher) super.visitVarInsn(opcode, var);
        else if (ByteCodeUtils.isLocalsStore(opcode) && var > this.locals) this.locals = var;
    }

    public abstract void visitInsn();
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/methodadapter/core/CoreMixinPatternMatcher.kt
package net.yakclient.mixin.base.internal.methodadapter.core

import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils
import net.yakclient.mixin.base.internal.bytecode.ByteCodeUtils.isLocalsStore
import net.yakclient.mixin.base.internal.instruction.Instruction
import net.yakclient.mixin.base.internal.instruction.core.CoreInsnExecutor
import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher
import org.objectweb.asm.MethodVisitor

abstract class CoreMixinPatternMatcher(visitor: MethodVisitor, private val instructions: Instruction) :
    MethodVisitor(ByteCodeUtils.ASM_VERSION, visitor), MixinPatternMatcher {
    var state = NOT_MATCHED
//    private var locals = 1
    fun executeInsn() {
        this.executeInsn(instructions)
    }

    private fun executeInsn(insn: Instruction?) {
        if (mv is CoreMixinPatternMatcher) (mv as CoreMixinPatternMatcher).executeInsn(insn) else CoreInsnExecutor(
            insn!!, mv
        ).execute()
    }

    override fun visitInsn(opcode: Int) {
        this.visitInsn()
        super.visitInsn(opcode)
    }

    override fun visitIntInsn(opcode: Int, operand: Int) {
        this.visitInsn()
        super.visitIntInsn(opcode, operand)
    }

    override fun visitVarInsn(opcode: Int, `var`: Int) {
        if (mv is CoreMixinPatternMatcher) super.visitVarInsn(
            opcode,
            `var`
        )
    }

    abstract fun visitInsn()

    companion object {
        const val NOT_MATCHED = 0
    }
}